message_passing/message_handler_waiting_compensation_cleanup/workflows.py

```diff
@@ -0,0 +1,121 @@
+    OnWorkflowExitAction,
+    UpdateInput,
+    WorkflowExitType,
+    WorkflowInput,
+)
+from message_passing.message_handler_waiting_compensation_cleanup.activities import (
+    my_activity,
+)
+
+
+@workflow.defn
+class MyWorkflow:
+    """
+    This Workflow upholds the following recommended practices:
+
+    1. The main workflow method ensures that all signal and update handlers are
+       finished before a successful return, and on failure, cancellation, and
+       continue-as-new.
+    2. The update handler performs any necessary compensation/cleanup when the
+       workflow is cancelled, fails, or continues-as-new.
+    """
+
+    def __init__(self) -> None:
+        self.workflow_exit_exception: asyncio.Future[BaseException] = asyncio.Future()
+
+    @workflow.run
+    async def run(self, input: WorkflowInput) -> str:
+        try:
+            # 👉 Use this `try...except` style, instead of waiting for message
+            # handlers to finish in a `finally` block. The reason is that other
+            # exception types will cause a Workflow Task failure, in which case
+            # we do *not* want to wait for message handlers to finish.
+            result = await self._run(input)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            return result
+        except (
+            asyncio.CancelledError,
+            workflow.ContinueAsNewError,
+            exceptions.FailureError,
+        ) as exc:
+            self.workflow_exit_exception.set_result(exc)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            raise exc
+
+    @workflow.update
+    async def my_update(self, input: UpdateInput) -> str:
+        """
+        This update handler demonstrates how to handle the situation where the
+        main Workflow method exits prematurely. In that case we perform
+        compensation/cleanup, and fail the Update. The Update caller will get a
+        WorkflowUpdateFailedError.
+        """
+        # Coroutines must be wrapped in tasks in order to use workflow.wait.
+        update_task = asyncio.Task(self._my_update())
+        # 👉 Always use `workflow.wait` instead of `asyncio.wait` in Workflow
+        # code: asyncio's version is non-deterministic.
+        first_completed, _ = await workflow.wait(  # type: ignore
+            [update_task, self.workflow_exit_exception],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+        # 👉 It's possible that the update completed and the workflow exited
+        # prematurely in the same tick of the event loop. If the Update has
+        # completed, return the Update result to the caller, whether or not the
+        # Workflow is exiting.
+        if (
+            update_task in first_completed
+            or input.on_premature_workflow_exit == OnWorkflowExitAction.CONTINUE
+        ):
+            return await update_task
+        else:
+            await self._my_update_compensation_and_cleanup()
```

Comment:
May want to demonstrate running an activity in the cleanup since that's the most common approach. May also want to cancel the update's activity first (whether you want to wait for update cancellation is up to you).

Also, one of the problems that plagues compensation writers is what to do with an error on compensation. Will leave it up to you, but at the least if we want to leave it like this, should acknowledge in a comment above this that we are intentionally letting raised compensation exceptions swallow the existing one.

Reply:
Agreed, done.